File: testdata/fixtures/java-enterprise/api/src/main/java/com/example/controllers/UserController.java
package com.example.controllers;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * User controller — handles user CRUD operations.
 * Known complexity: createUser=6, getUser=2, updateUser=5, handleBulkOperation=10
 */
public class UserController {

    // complexity: 1 (base) + 5 decision points = 6
    public Map<String, Object> createUser(Map<String, Object> request) {
        String email = (String) request.get("email");
        String name = (String) request.get("name");

        if (email == null || email.isEmpty()) {
            return Map.of("error", "Email required");
        }

        if (!email.contains("@")) {
            return Map.of("error", "Invalid email");
        }

        if (name == null || name.length() < 2) {
            return Map.of("error", "Name too short");
        }

        String role = (String) request.getOrDefault("role", "user");
        if (!List.of("user", "admin", "moderator").contains(role)) {
            return Map.of("error", "Invalid role");
        }

        try {
            return Map.of("id", "usr_123", "email", email, "name", name, "role", role);
        } catch (Exception e) {
            return Map.of("error", "Creation failed: " + e.getMessage());
        }
    }

    // complexity: 1 (base) + 1 decision point = 2
    public Map<String, Object> getUser(String userId) {
        Optional<Map<String, Object>> user = findById(userId);
        if (user.isEmpty()) {
            return Map.of("error", "User not found");
        }
        return user.get();
    }

    // complexity: 1 (base) + 4 decision points = 5
    public Map<String, Object> updateUser(String userId, Map<String, Object> updates) {
        Map<String, Object> user = getUser(userId);
        if (user.containsKey("error")) {
            return user;
        }

        if (updates.containsKey("email")) {
            String newEmail = (String) updates.get("email");
            if (newEmail == null || !newEmail.contains("@")) {
                return Map.of("error", "Invalid email");
            }
        }

        if (updates.containsKey("role")) {
            String newRole = (String) updates.get("role");
            if (!List.of("user", "admin").contains(newRole)) {
                return Map.of("error", "Invalid role");
            }
        }

        if (updates.containsKey("name")) {
            String newName = (String) updates.get("name");
            if (newName.length() < 2) {
                return Map.of("error", "Name too short");
            }
        }

        return Map.of("id", userId, "status", "updated");
    }

    // complexity: 1 (base) + 9 decision points = 10
    public Map<String, Object> handleBulkOperation(String operation, List<Map<String, Object>> items) {
        if (items == null || items.isEmpty()) {
            return Map.of("error", "No items provided");
        }

        if (items.size() > 100) {
            return Map.of("error", "Bulk limit exceeded");
        }

        int success = 0;
        int failed = 0;

        switch (operation) {
            case "create":
                for (Map<String, Object> item : items) {
                    Map<String, Object> result = createUser(item);
                    if (result.containsKey("error")) {
                        failed++;
                    } else {
                        success++;
                    }
                }
                break;
            case "delete":
                for (Map<String, Object> item : items) {
                    String id = (String) item.get("id");
                    if (id != null) {
                        success++;
                    } else {
                        failed++;
                    }
                }
                break;
            case "deactivate":
                for (Map<String, Object> item : items) {
                    String id = (String) item.get("id");
                    if (id != null) {
                        success++;
                    } else {
                        failed++;
                    }
                }
                break;
            default:
                return Map.of("error", "Unknown operation: " + operation);
        }

        return Map.of("success", success, "failed", failed, "total", items.size());
    }

    private Optional<Map<String, Object>> findById(String id) {
        return Optional.empty();
    }
}

File: testdata/fixtures/neglected-project/src/controllers/UserController.php
<?php
// UserController — legacy controller with high complexity
// Known complexity: handleRequest=15, processForm=12, validateInput=8

namespace App\Controllers;

class UserController
{
    // complexity: 1 (base) + 14 decision points = 15
    public function handleRequest($request, $response)
    {
        $method = $request->getMethod();
        $path = $request->getPath();

        if ($method === 'GET') {
            if ($path === '/users') {
                $page = $request->get('page', 1);
                $limit = $request->get('limit', 20);
                if ($limit > 100) {
                    $limit = 100;
                }
                $users = $this->getUsers($page, $limit);
                if (empty($users)) {
                    return $response->json(['error' => 'No users found'], 404);
                }
                return $response->json($users);
            } elseif (preg_match('/\/users\/(\d+)/', $path, $matches)) {
                $user = $this->getUser($matches[1]);
                if (!$user) {
                    return $response->json(['error' => 'User not found'], 404);
                }
                return $response->json($user);
            }
        } elseif ($method === 'POST') {
            if ($path === '/users') {
                $data = $request->getBody();
                $errors = $this->validateInput($data);
                if (!empty($errors)) {
                    return $response->json(['errors' => $errors], 422);
                }
                $user = $this->createUser($data);
                if (!$user) {
                    return $response->json(['error' => 'Failed to create user'], 500);
                }
                return $response->json($user, 201);
            }
        } elseif ($method === 'DELETE') {
            if (preg_match('/\/users\/(\d+)/', $path, $matches)) {
                $this->deleteUser($matches[1]);
                return $response->json(['status' => 'deleted']);
            }
        }

        return $response->json(['error' => 'Not found'], 404);
    }

    // complexity: 1 (base) + 11 decision points = 12
    public function processForm($data)
    {
        $result = [];

        if (!isset($data['name']) || empty($data['name'])) {
            $result['errors'][] = 'Name is required';
        } elseif (strlen($data['name']) < 2) {
            $result['errors'][] = 'Name too short';
        } elseif (strlen($data['name']) > 100) {
            $result['errors'][] = 'Name too long';
        }

        if (isset($data['email'])) {
            if (!filter_var($data['email'], FILTER_VALIDATE_EMAIL)) {
                $result['errors'][] = 'Invalid email';
            } else {
                $existing = $this->findByEmail($data['email']);
                if ($existing && $existing['id'] !== ($data['id'] ?? null)) {
                    $result['errors'][] = 'Email already taken';
                }
            }
        }

        if (isset($data['age'])) {
            if (!is_numeric($data['age'])) {
                $result['errors'][] = 'Age must be numeric';
            } elseif ($data['age'] < 0 || $data['age'] > 150) {
                $result['errors'][] = 'Invalid age';
            }
        }

        if (empty($result['errors'])) {
            $result['valid'] = true;
        }

        return $result;
    }

    // complexity: 1 (base) + 7 decision points = 8
    public function validateInput($data)
    {
        $errors = [];

        if (!is_array($data)) {
            return ['Invalid input format'];
        }

        foreach (['name', 'email'] as $required) {
            if (!isset($data[$required]) || empty($data[$required])) {
                $errors[] = "$required is required";
            }
        }

        if (isset($data['email']) && !filter_var($data['email'], FILTER_VALIDATE_EMAIL)) {
            $errors[] = 'Invalid email format';
        }

        if (isset($data['role'])) {
            $validRoles = ['user', 'admin', 'moderator'];
            if (!in_array($data['role'], $validRoles)) {
                $errors[] = 'Invalid role';
            }
        }

        if (isset($data['password'])) {
            if (strlen($data['password']) < 8) {
                $errors[] = 'Password too short';
            }
            if (!preg_match('/[A-Z]/', $data['password'])) {
                $errors[] = 'Password needs uppercase';
            }
        }

        return $errors;
    }

    private function getUsers($page, $limit) { return []; }
    private function getUser($id) { return null; }
    private function createUser($data) { return $data; }
    private function deleteUser($id) { }
    private function findByEmail($email) { return null; }
}
